_posts/2018-10-02-auto-tune-all.md

```diff
@@ -0,0 +1,198 @@
+we are able to generate kernels that are comparable to heavily hand-optimized libraries,
+and obtain the state-of-the-art inference performance on hardware platforms including
+ARM CPUs, Intel CPUs, Mali GPUs, NVIIDA GPUs and AMD GPUs.
+
+In this blog post, I will show the workflow of automatic kernel optimization in TVM compiler stack and 
+benchmark results on several hardware platforms.
+
+# System Overview
+
+{:center: style="text-align: center"}
+![image](/images/autotune-all/circle.png){: width="35%"}
+{:center}
+<center> Figure 1. System Overview </center> <p></p>
+
+Kernel optimization in TVM is done in an iterative loop fashion.
+As shown in Figure 1, the automatic kernel optimization takes a neural network (typically in computational graph representation)
+from frontend frameworks as input, and generates kernels for all the operators in this network.
+
+The inner loop consists of scalable RPC runtime, machine learning based tuners and a tensor compiler.
+In each round of the loop, the tuner picks a batch of promising candidate kernel implementations from a large search space,
+and measures them on real hardware. Then the tuner gets the profiling results. These profiling results are used as training 
+data to fit a prediction model. After fitting the prediction model, the tuner picks the next promising candidates according to the predictions,
+and the loop continues. By this way, we search for fast kernels iteratively.
+
+The below figure shows the difference between traditional auto-tuning and
+AutoTVM. You can refer to our [paper](https://arxiv.org/abs/1805.08166)
+for more details.
+
+{:center: style="text-align: center"}
+![image](/images/autotune-all/autotvm.png){: width="50%"}
+{:center}
+<center> Figure 2. Traditional Auto-tuning vs AutoTVM </center> <p></p>
+
+## Begin Tuning
+For demonstration, we run our optimization for resnet-18 on RK3399, an ARM development board.
+The detailed instructions are omitted due to space limit of a blog post.
+Step-by-step tutorials are available for
+[ARM CPU](https://docs.tvm.ai/tutorials/autotvm/tune_nnvm_arm.html),
+[Mobile GPU](https://docs.tvm.ai/tutorials/autotvm/tune_nnvm_mobile_gpu.html),
+[NVIDIA/AMD GPU](https://docs.tvm.ai/tutorials/autotvm/tune_nnvm_cuda.html).
+
+First we get a pre-trained model from MXNet model zoo, and extract tuning tasks from it.
+```python
+from mxnet.gluon.model_zoo.vision import get_model
+
+block = get_model('resnet18_v1', pretrained=True)
+net, params = nnvm.frontend.from_mxnet(block)
+
+tasks = autotvm.extract_from_graph(net)
+tune_tasks(tasks, **tuning_option)
+```
+There are 12 different conv2d layers in resnet-18, so we launched 12 tuning tasks.
+For each of them, the tuner made several hundreds of trials and picked the best one.
+After finishing all tuning tasks, we compiled the whole network and generated a single deployable minimal library.
```

Comment:
... we compile the whole network and generate a single deployable minimal library.